src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,291 @@
+                asyncListener,
+                _cancellationToken);
+
+            _filterAndSortDataModelQueue = new AsyncBatchingWorkQueue<bool, DocumentSymbolDataModel?>(
+                DelayTimeSpan.NearImmediate,
+                FilterAndSortDataModelAsync,
+                EqualityComparer<bool>.Default,
+                asyncListener,
+                _cancellationToken);
+
+            _highlightExpandAndPresentItemsQueue = new AsyncBatchingWorkQueue<ExpansionOption>(
+                DelayTimeSpan.NearImmediate,
+                HighlightExpandAndPresentItemsAsync,
+                asyncListener,
+                _cancellationToken);
+
+            // We don't think the shell is initialized lazily, so we'll Debug.Fail(), but if it was we'd still
+            // see the view created later so this will still function.
+            if (ErrorHandler.Failed(codeWindow.GetPrimaryView(out var primaryTextView)))
+                Debug.Fail("GetPrimaryView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Failed(StartTrackingView(primaryTextView)))
+                Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Succeeded(codeWindow.GetSecondaryView(out var secondaryTextView)))
+            {
+                if (ErrorHandler.Failed(StartTrackingView(secondaryTextView)))
+                    Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+            }
+
+            var subjectBuffer = _trackedTextViews[primaryTextView].TextBuffer;
+            _textViewEventSource = new CompilationAvailableTaggerEventSource(
+                subjectBuffer,
+                _asyncListener,
+                // Any time an edit happens, recompute as the document symbols may have changed.
+                TaggerEventSources.OnTextChanged(subjectBuffer),
+                // Switching what is the active context may change the document symbols.
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer),
+                // Many workspace changes may need us to change the document symbols (like options changing, or project renaming).
+                TaggerEventSources.OnWorkspaceChanged(subjectBuffer, _asyncListener),
+                // Once we hook this buffer up to the workspace, then we can start computing the document symbols.
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer));
+
+            _textViewEventSource.Changed += OnEventSourceChanged;
+            _textViewEventSource.Connect();
+            _codeWindowEventsSink = ComEventSink.Advise<IVsCodeWindowEvents>(codeWindow, this);
+            StartComputeDataModelTask();
+        }
+
+        public void Dispose()
+        {
+            _codeWindowEventsSink.Unadvise();
+            _textViewEventSource.Changed -= OnEventSourceChanged;
+            _textViewEventSource.Disconnect();
```

Comment:
i would recommend going in reverse order here.  so Unadvise->Disconnect->EventUnsubscribe.

Reply:
@CyrusNajmabadi: hmm, event unsubscribe is legal after disposing an object? I would have assumed that the underlying object throwing ObjectDisposed was allowed there?

Reply:
That's a reasonable concern. Let's leave as is then!